futures-core/src/executor.rs

```diff
@@ -1,62 +1,61 @@
+    ///
+    /// # Returns
     ///
-    /// A *task* is a `()`-producing future that runs at the top level, and will
-    /// be `poll`ed until completion. It's also the unit at which wake-up
-    /// notifications occur. Executors, such as thread pools, allow tasks to be
-    /// spawned and are responsible for putting tasks onto ready queues when
-    /// they are woken up, and polling them when they are ready.
-    pub trait Executor {
-        /// Spawn the given task, polling it until completion.
-        ///
-        /// Tasks must be infallible, as the type suggests; it is the
-        /// client's reponsibility to route any errors elsewhere via a channel
-        /// or some other means of communication.
-        ///
-        /// # Errors
-        ///
-        /// The executor may be unable to spawn tasks, either because it has
-        /// been shut down or is resource-constrained.
-        fn spawn(&mut self, f: Box<Future<Item = (), Error = Never> + Send>) -> Result<(), SpawnError>;
-
-        /// Determine whether the executor is able to spawn new tasks.
-        ///
-        /// # Returns
-        ///
-        /// An `Ok` return means the executor is *likely* (but not guaranteed)
-        /// to accept a subsequent spawn attempt. Likewise, an `Err` return
-        /// means that `spawn` is likely, but not guaranteed, to yield an error.
-        fn status(&self) -> Result<(), SpawnError> {
-            Ok(())
-        }
-
-        // TODO: downcasting hooks
+    /// An `Ok` return means the executor is *likely* (but not guaranteed)
+    /// to accept a subsequent spawn attempt. Likewise, an `Err` return
+    /// means that `spawn` is likely, but not guaranteed, to yield an error.
+    fn status(&self) -> Result<(), SpawnErrorKind> {
+        Ok(())
     }
 
-    /// Provides the reason that an executor was unable to spawn.
-    #[derive(Debug)]
-    pub struct SpawnError {
-        _a: ()
-    }
+    // TODO: downcasting hooks
+}
 
-    impl SpawnError {
-        /// Spawning is failing because the executor has been shut down.
-        pub fn shutdown() -> SpawnError {
-            SpawnError { _a: () }
-        }
+/// Provides the reason that an executor was unable to spawn.
+#[derive(Debug)]
+pub struct SpawnErrorKind {
+    _a: ()
+}
+
+impl SpawnErrorKind {
+    /// Spawning is failing because the executor has been shut down.
+    pub fn shutdown() -> SpawnErrorKind {
+        SpawnErrorKind { _a: () }
+    }
 
-        /// Check whether this error is the `shutdown` error.
-        pub fn is_shutdown() -> bool {
-            true
-        }
+    /// Check whether this error is the `shutdown` error.
+    pub fn is_shutdown() -> bool {
+        true
     }
 }
 
-#[cfg(not(feature = "std"))]
-pub(crate) trait Executor {}
+/// The result of a failed spawn
+#[derive(Debug)]
+pub struct SpawnError<T> {
+    /// The kind of error
+    pub kind: SpawnErrorKind,
+
+    /// The task for which spawning was attempted
+    pub task: T,
```

Comment:
As I mentioned on IRC, I don't know that we want to commit ourselves to handing back the task, since this won't be possible when we move to unsized rvalues (at least not with some novel new stack-DST-returning ABI).

Reply:
So, interesting thing: this does *not* make such a commitment for the unsized-rvalue spawning, which will be a distinct method (called `spawn`) once it's possible.

In that world, `spawn_obj` will remain available though largely unused -- but perhaps one use-case for it is precisely when you need to be able to get your task back, at the cost of needing to pre-allocate it.

In that case the eventual `spawn` method would return a `SpawnErrorKind` rather than `SpawnError`.

wdyt? if that seems worthwhile, I should probably rejigger things to get rid of the type parameter and call it `SpawnObjError`.

Reply:
yeah, we could add a different method later on, but it seems like it sets up the expectation that users might want their task back, and the name `SpawnError` suggests that it would be the error for the `spawn` method. I think it's fine to keep it this way with some renaming etc.

Reply:
Done